### PR TITLE
fix: reject full-chain symlink escapes in file directives (#67)

### LINKED ISSUE #67: follow-up: Symlink full-chain resolution via EvalSymlinks (deferred from #52)
Filed per v0.33.0 spec § Follow-up issues #3.
See: https://github.com/2389-research/dippin-lang/blob/main/docs/superpowers/specs/2026-05-27-issue-52-command-file-design.md

**Symlink full-chain resolution.** v1 uses `os.Lstat` to reject symlinks at the directive's path. Full `EvalSymlinks` resolution (catching symlinks deeper in the resolved path) is deferred. v1 only rejects symlinks at the directive's path. A deeper symlink (in a parent directory that resolves elsewhere) could still escape. Tighten if the threat model requires.

### PR DESCRIPTION
**Pull request by clintecker**
Closes #67.

## Problem

The file-directive path-security resolver (`parser/resolve.go`) had a path-traversal gap. It enforced containment with:
1. **Lexical** parent-escape detection (`filepath.Rel`) in `safeResolve` — never resolves symlinks.
2. A **leaf-only** symlink check (`os.Lstat` on the resolved path) in `checkFileInfo`.

A symlinked **parent directory** defeats both — e.g. `baseDir/sub -> /etc`, then `command_file: sub/passwd`:
- The leaf (`passwd`) is a real file, so `Lstat` sees no symlink.
- Lexical `Rel(baseDir, baseDir/sub/passwd)` = `sub/passwd` — no `..`, so it passes.

Result: `dippin validate` / `dippin pack` would read files outside the `.dip` source directory.

## Fix

Add `checkContainment`: resolve the file's **parent directory** through `filepath.EvalSymlinks` and reject if its real location escapes the real `baseDir`. Resolving the parent (not the leaf) leaves the existing leaf-symlink rejection intact, so error messages and behavior for direct leaf symlinks are unchanged.

Refactored out two shared helpers to stay within the repo's complexity caps (cyclomatic ≤5 / cognitive ≤7):
- `escapesBase` — the `Rel`-based containment predicate, now shared by the lexical and full-chain checks.
- `pathErr` — the user-path-only error mapping (no resolved-absolute-path leak), shared by stat/read/containment.

## Tests (TDD — test written and watched fail first)

- `RejectsSymlinkedParentDir` — the escape is now rejected (`resolves outside source directory`).
- `AllowsSymlinkedParentDirInsideBase` — guard: symlinked dirs that stay **inside** `baseDir` still load (no over-blocking).
- All existing resolver tests (leaf-symlink rejection, parent escape, oversize, missing-file, agent slots) still pass.

`just check` green: 0 lint issues, complexity OK, race tests pass, examples validate.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/2389-research/codesmith/dippin-lang/pr/77"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1782592674&installation_id=131176874&pr_number=77&repository=2389-research%2Fdippin-lang&return_to=https%3A%2F%2Fgithub.com%2F2389-research%2Fdippin-lang%2Fpull%2F77&signature=7fe24c7f31d0f92637d8bd7c7461ae618a4b491facec5383ad9cd8356bf67edf"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Symlink handling tightened: containment is now checked earlier in the resolution flow to prevent path escapes.
  * Error reporting consolidated so filesystem failures are shown as user-path-only diagnostics for clearer, less noisy messages.

* **Tests**
  * Added tests covering symlink edge cases involving parent directories, relative bases, and internal symlinks to verify containment behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/2389-research/dippin-lang/pull/77?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: parser/resolve.go [hunk 1]
```diff
@@ -122,28 +122,66 @@ func statDirectiveFile(p, resolved string) (os.FileInfo, error) {
...
+// does lexical containment and checkFileInfo's Lstat only inspects the leaf, so
+// a symlinked *parent* directory (e.g. baseDir/sub -> /etc, then sub/passwd)
+// defeats both. Resolving the parent — not the leaf — leaves the separate
+// leaf-symlink rejection in checkFileInfo intact. (#67)
+func checkContainment(baseDir, resolved, p string) error {
+	realBase, err := filepath.EvalSymlinks(baseDir)
+	if err != nil {
+		return pathErr(p, err, "stat")
+	}
+	realParent, err := filepath.EvalSymlinks(filepath.Dir(resolved))
```
[the remark this answers is not in the document]
Comment (clintecker) [reply parser/resolve.go]: Verified and fixed in 8b64666. `filepath.Rel(".", "/abs")` does error, so a relative baseDir plus an internal symlink with an absolute target was falsely rejected. Added a `realPath` helper (EvalSymlinks + filepath.Abs) so both base and parent are absolute before `escapesBase`, and a regression test covering a relative baseDir (`dippin validate workflow.dip`).

File: parser/resolve.go [hunk 2]
```diff
@@ -122,28 +122,66 @@ func statDirectiveFile(p, resolved string) (os.FileInfo, error) {
...
+// and rejects the path if its real location escapes baseDir. safeResolve only
+// does lexical containment and checkFileInfo's Lstat only inspects the leaf, so
+// a symlinked *parent* directory (e.g. baseDir/sub -> /etc, then sub/passwd)
+// defeats both. Resolving the parent — not the leaf — leaves the separate
+// leaf-symlink rejection in checkFileInfo intact. (#67)
+func checkContainment(baseDir, resolved, p string) error {
+	realBase, err := filepath.EvalSymlinks(baseDir)
+	if err != nil {
+		return pathErr(p, err, "stat")
+	}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply parser/resolve.go:166]: Leaving as-is. `EvalSymlinks(baseDir)` runs once per directive, and workflows carry a handful of directives — the cost is dominated by the file reads themselves. Hoisting the real-base computation up to `ResolveFileDirectives` would thread it through `loadDirectiveInto` → `loadDirectiveFile` → `checkContainment`, which isn't worth the signature churn at this scale.

**Comment by clintecker** [comment]
Re: CodeRabbit's outside-diff TOCTOU note on `parser/resolve.go` (check-then-read race).

Valid finding, but intentionally out of scope for this PR. #67 closes the *static/structural* symlinked-parent escape, which is the exploitable gap for a malicious or mistaken `.dip` + filesystem layout. A TOCTOU race requires an attacker who can mutate the filesystem *during* a validate/pack run — for dippin's primary model (an author running the tool on their own `.dip` in their own directory) that adds little, since the author already controls the directory. It matters as defense-in-depth for untrusted-`.dip`-in-CI scenarios, and the fix (fd-based `openat`/`O_NOFOLLOW` + `fstat` + read-from-fd) is platform-specific.

Tracked as #79. Keeping #67 scoped to the structural escape.
